backend/src/main/java/com/empire/World.java

```diff
@@ -962,6 +967,82 @@ void orderOverrides() {
+					boolean templeBonus = true;
+					for (Region r : region.getNeighbors(World.this)) if (r.isLand() && (r.religion != region.religion || r.culture != region.culture)) templeBonus = false;
+					if (templeBonus) costMod -= 1;
+				}
+				ct = Construction.makeTemple(ideo, Math.max(0, Constants.baseCostTemple * costMod));
+			} else if (action.contains("Fortifications")) {
+				if (Ideology.FLAME_OF_KITH == region.religion) costMod -= 1;
+				ct = Construction.makeFortifications(Math.max(0, Constants.baseCostFortifications * costMod));
+			} else {
+				return;
+			}
+			double cost = ct.originalCost;
+			if (cost <= nation.gold) {
+				nation.gold -= cost;
+				incomeSources.getOrDefault(kingdom, new Budget()).spentConstruction += cost;
+				region.constructions.add(ct);
+				if (ct.type == Construction.Type.TEMPLE) {
+					Ideology r = region.religion;
+					region.setReligion(ct.religion, World.this);
+					if (r != Ideology.VESSEL_OF_FAITH && region.religion == Ideology.VESSEL_OF_FAITH) {
+						for (String k : kingdoms.keySet()) if (Ideology.VESSEL_OF_FAITH == NationData.getStateReligion(k, World.this)) {
+							for (Region rr : regions) if (k.equals(rr.getKingdom())) rr.unrestPopular = Math.max(0, rr.unrestPopular - .1);
+						}
+					}
+					if (nation.hasTag(NationData.Tag.MYSTICAL)) region.unrestPopular = Math.max(0, region.unrestPopular - .1);
+					if (church.hasDoctrine(Church.Doctrine.ANTIECUMENISM) && ct.religion.religion != Religion.IRUHAN) nation.goodwill += Constants.antiecumenismConstructionOpinion;
+					if (church.hasDoctrine(Church.Doctrine.ANTISCHISMATICISM) && ct.religion == Ideology.VESSEL_OF_FAITH) nation.goodwill += Constants.antischismaticismConstructionOpinion;
+					if (church.hasDoctrine(Church.Doctrine.WORKS_OF_IRUHAN) && ct.religion.religion == Religion.IRUHAN) nation.goodwill += Constants.worksOfIruhanConstructionOpinion;
+				}
+				if (ct.type == Construction.Type.SHIPYARD && nation.hasTag(NationData.Tag.SHIP_BUILDING)) buildShips(kingdom, regions.indexOf(region), Constants.numShipsBuiltPerShipyard * Constants.shipBuildingTraitWeeksProduction);
+				builds.add(region);
+				if (ct.type == Construction.Type.TEMPLE) templeBuilds.add(region);
+				if (ct.type == Construction.Type.FORTIFICATIONS && nation.hasTag(NationData.Tag.DEFENSIVE)) {
+					region.constructions.add(Construction.makeFortifications(0));
+				}
+			} else {
+				notifications.add(new Notification(kingdom, "Construction Failed", "We did not have the " + Math.round(cost) + "gold necessary to construct as ordered in " + region.name + "."));
+			}
+		}
+
+		void nobleActions() {
+			for (int rid = 0; rid < regions.size(); rid++) {
+				Region r = regions.get(rid);
+				if (!r.hasNoble()) continue;
+				if (r.noble.unrest > 0.5) continue;
+				String action = orders.getOrDefault(r.getKingdom(), new HashMap<String, String>()).getOrDefault("action_noble_" + rid, "Relax");
+				if (action.startsWith("Build ")) {
```

Comment:
The translation of orders into Java objects has been on my mind a bunch.  I was thinking some kind of static factory with lots of methods like orderToNobleAction, orderToArmyAction, etc. might be a good place to start capturing all of those somewhat brittle string matching rules and move towards a better way of doing the translation.

Reply:
Yeah. I started playing with that a little bit by introducing an enum for certain noble orders. My thinking was that every object (character, army, noble, etc) could hold on transiently to its order, which could simplify some of the tracking that World.Advancer is doing. If we had enum types, we could straight up deserialize them using GSON or something, and skip the string representations. However, the part I struggled with was that orders have arguments (e.g. travel to |x|), which is hard to do with an enum. Still looking for a good solution; I agree with you that this approach is brittle, though. (It's also hard to internationalize, not that I'm looking to do that any time soon.)